File: nko-admin-panel/src/app/services/statistics.service.ts
// statistics.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class StatisticsService {
  private apiUrl = environment.apiUrl + environment.endpoints.statistics;

  constructor(private http: HttpClient) { }

  getCompleteStatistics(includeAdvanced: boolean = false): Observable<StatisticsResponse> {
    return this.http.get<StatisticsResponse>(`${this.apiUrl}/?include_advanced=${includeAdvanced}`);
  }

  getBaseStatistics(): Observable<BaseStatsResponse> {
    return this.http.get<BaseStatsResponse>(`${this.apiUrl}/base`);
  }

  getPopularCategories(limit: number = 10): Observable<CategoryStats[]> {
    return this.http.get<CategoryStats[]>(`${this.apiUrl}/popular/categories?limit=${limit}`);
  }

  getPopularItems(limit: number = 10): Observable<PopularItem[]> {
    return this.http.get<PopularItem[]>(`${this.apiUrl}/popular/items?limit=${limit}`);
  }

  getAdvancedStatistics(): Observable<AdvancedStats> {
    return this.http.get<AdvancedStats>(`${this.apiUrl}/advanced`);
  }
}

export interface BaseStatsResponse {
  total_users: number;
  total_lessors: number;
  total_equipment: number;
  approved_bookings: number;
  rejected_bookings: number;
}

export interface CategoryStats {
  category_id: number;
  category_name: string | null;
  equipment_count: number;
  booking_count: number;
}

export interface PopularItem {
  equipment_id: number;
  equipment_name: string | null;
  booking_count: number;
  total_quantity_booked: number;
}

export interface AdvancedStats {
  avg_rental_duration_days: number | null;
  active_users_count: number;
  total_rental_days: number;
}

export interface StatisticsResponse {
  base_stats: BaseStatsResponse;
  popular_categories: CategoryStats[];
  popular_items: PopularItem[];
  advanced_stats?: AdvancedStats;
}

File: app/services/statistics_service.py
# app/services/statistics_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Dict, Any, List, Optional
from app.repositories.statistics_repository import StatisticsRepository
from app.api.schemas.statistics_schema import (
    BaseStatsResponse, CategoryStats, PopularItem, 
    AdvancedStats, StatisticsResponse
)

class StatisticsService:
    
    def __init__(self):
        self.repository = StatisticsRepository()
    
    async def get_base_statistics(self, session: AsyncSession) -> BaseStatsResponse:
        """Получить базовую статистику"""
        base_stats = await self.repository.get_base_stats(session)
        return BaseStatsResponse(**base_stats)
    
    async def get_popular_categories(self, session: AsyncSession, limit: int = 10) -> List[CategoryStats]:
        """Получить популярные категории"""
        categories_data = await self.repository.get_popular_categories(session, limit)
        
        # Здесь можно добавить логику для получения названий категорий
        # если у вас есть соответствующая модель
        return [
            CategoryStats(
                category_id=cat["category_id"],
                category_name=cat['category_name'],
                equipment_count=cat["equipment_count"],
                booking_count=cat["booking_count"]
            )
            for cat in categories_data
        ]
    
    async def get_popular_items(self, session: AsyncSession, limit: int = 10) -> List[PopularItem]:
        """Получить популярное оборудование"""
        items_data = await self.repository.get_popular_items(session, limit)
        
        return [
            PopularItem(
                equipment_id=item["equipment_id"],
                equipment_name=item["equipment_name"],
                booking_count=item["booking_count"],
                total_quantity_booked=item["total_quantity_booked"]
            )
            for item in items_data
        ]
    
    async def get_advanced_statistics(self, session: AsyncSession) -> AdvancedStats:
        """Получить расширенную статистику"""
        advanced_stats = await self.repository.get_advanced_stats(session)
        return AdvancedStats(**advanced_stats)
    
    async def get_complete_statistics(
        self, 
        session: AsyncSession, 
        include_advanced: bool = False
    ) -> StatisticsResponse:
        """Получить полную статистику"""
        base_stats = await self.get_base_statistics(session)
        popular_categories = await self.get_popular_categories(session)
        popular_items = await self.get_popular_items(session)
        
        advanced_stats = None
        if include_advanced:
            advanced_stats = await self.get_advanced_statistics(session)
        
        return StatisticsResponse(
            base_stats=base_stats,
            popular_categories=popular_categories,
            popular_items=popular_items,
            advanced_stats=advanced_stats
        )